index.bs

```diff
@@ -12839,6 +12871,34 @@ running the algorithm for an {{AudioNode}}, using an <a>input
 buffer</a> and the value(s) of the {{AudioParam}}(s) of this
 {{AudioNode}} as the input for this algorithm.
 
+
+<h3 id="error-handling-on-a-running-audio-context">
+    Error Handling for System Audio Resources on a running AudioContext</h3>
+
+In the event of an audio system resource error, the <a>rendering thread</a>
+performs the following steps upon receiving the error:
+
+1. Set {{[[rendering thread state]]}} of the {{AudioContext}} to
+    "{{AudioContextState/suspended}}".
+
+1. <a href="https://html.spec.whatwg.org/multipage/media.html#queue-a-media-element-task">
+    Queue a media element task</a> to execute the following steps:
+
+    1. <a spec="dom" lt="fire an event">fire an event</a> named
+        {{AudioContext/error}} at the associated {{AudioContext}}.
+    
+    1. If the {{BaseAudioContext/state}} attribute of the {{AudioContext}} is
+        not already "{{AudioContextState/suspended}}":
```

Comment:
Per our brief discussion; I suggest changing this line to:

`"{{AudioContextState/running}}":`

Reply:
Correct.

That brings another question: do we want to fire `onerror` even when the context is closed? What does that mean and how's that useful to developers?

I actually prefer not to do anything when the context is closed. The closed context should be of no use.

To summarize:
- If the context is suspended, don't change anything and just fire `onerror`.
- If the context is running, suspend the context and then fire `onerror`.
- if the context is closed, just return.

And yet another question - if the app tries to resume the suspended context (due to `onerror`) what should happen? Also do we have to change the previous sinkId (e.g. to a default device) because it was a problematic device?

cc @padenot

Reply:
Does that mean the `'error'` event will replace the `'sinkchange'` event proposed here (https://github.com/WebAudio/web-audio-api/issues/2532#issuecomment-1717151574)?

Reply:
@chrisguttandin If the group's answer to my last question above is yes, then yes.

statechange (suspended) -> error -> sinkchange

The flow makes sense, but perhaps too many events are triggered?

Reply:
If a developer wants to use a non-default sink, they should know what they are doing. They should know it may fail. I'd rather leave the decision of switching to a different sink (default or else) to the developers.

Reply:
I'd like to not have `"error"` fire in nominal scenario. `"statechange"` is for state changes that aren't an error.

Reply:
I am okay with that. To summarize:

When the running AudioContext encounters an error from the renderer, it suspends the context (but with no statechange event) and fires an error event.

WDYT?

Reply:
We usually pair handlers with an attribute. We could do the same as `HTMLMediaElement`, that's fairly flexible. It's probably that `setSinkId` + `resume` or just `setSinkId` would allow restarting a failed context.

Reply:
Regarding not firing `statechang` on an error:

At least the current [wording for onStateChange](https://webaudio.github.io/web-audio-api/#dom-baseaudiocontext-onstatechange) should be changed as well in this case. Maybe adding a note that the `onStateChange` handler won't be notified of state changes that are the result of an error?  Or perhaps say that the `onStateChange` handler is only notified in response to requested state changes or automatic playing?

I feel like it's probably simpler for implementers and developers to always fire `statechange` when the state changes. Are there use cases that become more difficult or impossible if we do it the simpler way?

Reply:
I'm not sure if we should blindly copy the `MediaRecorder` but it follows a different pattern. I always fires an `'error'` event followed by a `'stop'` event. The `'stop'` event would also fire when the `MediaRecorder` gets stopped without an error.

It's defined in step 15.3., step 15.4., and 15.5. of this algorithm: https://w3c.github.io/mediacapture-record/#dom-mediarecorder-start.

Reply:
Two comments above support firing both events. The benefits are a) it's consistent, b) easier to understand/implement, and b) developers can easily filter out events they do not need.

@padenot What's the benefit of not firing `statechange` upon the suspension caused by the system audio error?

Reply:
I think it's fine, in restrospect.

Reply:
Teleconf May02: Developer opinions are positive on both events being fired.

Reply:
@padenot Could you clarify your last comment? Are we supportive of firing both events (onerror & onstatechange) upon a system audio failure?

Reply:
Yes, both events.

Reply:
Great. Thanks for confirmation. Resolving.